### ISSUE TITLE
binActiveZ equivalent

### ISSUE THREAD

**Issue by xiaotiancd** [AngusG/tensorflow-xnor-bnn#1]
Hi there,
I am trying to use alexnet with xnor on tensorflow and comparing your code with the original paper and torch implementation by the author. Wondering how could I implement the equivalent BinActiveZ with tensorflow.  Would your bin_conv_net work, especially for the back-propagation to update the gradient? May you help to explain how the activation function is replaced, and how the gradient is updated based on your code, please. 

thank you!

**Comment by akhauriyash** [comment]
I cant answer this question, but can you tell me how you got the alexnet XNOR weights? The github gist link I saw was a torch .t7 file, and it seemed to be corrupt. If you can give me other link/format/share on drive a working version i'd be very grateful.
Please help! :octopus: 
Thanks!

**Comment by AngusG** [comment]
@xiaotiancd my `bin_conv_net` is slightly out of date with my private branch at the moment, but what I found works well is something like the following:

# Layer 1: full-precision weights and activations
Conv2D() 
ReLU()

# Layer 2:
BatchNorm() # mostly just to center output of ReLU around 0
BinActiv() # simple binarization works fine for MNIST
BinConv()
ReLU() #  +Pooling if desired

# Layer 3:
BatchNorm()
BinActiv()
BinConv()
ReLU() #  +Pooling if desired

...

For the BinActiveZ from the XNOR-Net paper, see the comments in `binary_conv_net.py`. In practice I found this wasn't necessary when scaling activations by a learned parameter for smaller scale problems at least. 

The gradient is updated by registering a straight through estimator for the `tf.sign()` op.
```
def quantize(self, x):
        with self.G.gradient_override_map({"Sign": "QuantizeGrad"}):
            return tf.sign(x)

@tf.RegisterGradient("QuantizeGrad")
def quantize_grad(op, grad):
    return tf.clip_by_value(tf.identity(grad), -1, 1)
```

**Comment by akhauriyash** [comment]
@xiaotiancd How did you port the torch weights to tensorflow?  I managed to inspect the Alexnet XNOR model file which is given in .t7 extension via lua torch which i foun on allenai github, but I do not know how to convert it to another format, like h5 weights/tensorflow model etc. Please help me out here. Really appreciate it! 🐙

**Comment by AngusG** [comment]
Closing due to inactivity

**Comment by xiaotiancd** [comment]
@akhauriyash i just re-train the data i need,e.g. imagenet
